router-tests/telemetry_test.go

```diff
@@ -2267,4 +2267,136 @@ Downstream errors:
+								require.NotEqual(t, "", r.Header.Get("x-datadog-parent-id"))
+								require.Equal(t, "1", r.Header.Get("x-datadog-sampling-priority"))
+								handler.ServeHTTP(w, r)
+							})
+						},
+					},
+				},
+			}, func(t *testing.T, xEnv *testenv.Environment) {
+				res := xEnv.MakeGraphQLRequestOK(testenv.GraphQLRequest{
+					Query: `query myQuery { employees { id } }`,
+					Header: map[string][]string{
+						"traceparent": {"00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203332-01"}, // 01 = sampled
+					},
+				})
+				require.JSONEq(t, employeesIDData, res.Body)
+			})
+		})
+
+		t.Run("Datadog headers correctly recognize sampling bit", func(t *testing.T) {
+			t.Parallel()
+
+			exporter := tracetest.NewInMemoryExporter(t)
+
+			testenv.Run(t, &testenv.Config{
+				TraceExporter:    exporter,
+				DatadogTelemetry: true,
+				Subgraphs: testenv.SubgraphsConfig{
+					Employees: testenv.SubgraphConfig{
+						Middleware: func(handler http.Handler) http.Handler {
+							return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+								require.Equal(t, datadogTraceId, r.Header.Get("x-datadog-trace-id"))
+								require.NotEqual(t, "", r.Header.Get("x-datadog-parent-id"))
+								require.Equal(t, "0", r.Header.Get("x-datadog-sampling-priority"))
+								handler.ServeHTTP(w, r)
+							})
+						},
+					},
+				},
+			}, func(t *testing.T, xEnv *testenv.Environment) {
+				res := xEnv.MakeGraphQLRequestOK(testenv.GraphQLRequest{
+					Query: `query myQuery { employees { id } }`,
+					Header: map[string][]string{
+						"traceparent": {"00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203332-00"}, // 01 = sampled
+					},
+				})
+				require.JSONEq(t, employeesIDData, res.Body)
+			})
+		})
+
+		t.Run("Correctly pass along Datadog headers", func(t *testing.T) {
+			t.Parallel()
+
+			exporter := tracetest.NewInMemoryExporter(t)
+
+			testenv.Run(t, &testenv.Config{
+				TraceExporter:    exporter,
+				DatadogTelemetry: true,
+				Subgraphs: testenv.SubgraphsConfig{
+					Employees: testenv.SubgraphConfig{
+						Middleware: func(handler http.Handler) http.Handler {
+							return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+								require.Equal(t, datadogTraceId, r.Header.Get("x-datadog-trace-id"))
+								require.NotEqual(t, "12345", r.Header.Get("x-datadog-parent-id"))
+								require.Equal(t, "1", r.Header.Get("x-datadog-sampling-priority"))
+								handler.ServeHTTP(w, r)
+							})
+						},
+					},
+				},
+			}, func(t *testing.T, xEnv *testenv.Environment) {
+				res := xEnv.MakeGraphQLRequestOK(testenv.GraphQLRequest{
+					Query: `query myQuery { employees { id } }`,
+					Header: map[string][]string{
+						"x-datadog-trace-id":          {datadogTraceId},
+						"x-datadog-parent-id":         {"12345"},
+						"x-datadog-sampling-priority": {"1"},
+					},
+				})
+				require.JSONEq(t, employeesIDData, res.Body)
```

Comment:
The only missing piece now, is to validate if the exported trace and span IDs are correct. You can test it very easily.

Reply:
@StarpTech what do you mean by that? What test were you thinking of?

Reply:
I'm able to validate that the trace ID is correct, but at some point in the process, it looks like the span ID gets overwritten so I'm having difficulty validating that. What do you recommend? I've tried stepping through with the debugger, I'm not sure how long to spend on it

Reply:
Hi @df-wg, I meant to validate the traceId and spanId exporter by the router. It's correct that a different span is generated because it's a different unit of work.

Reply:
@StarpTech I'm doing that here: https://github.com/wundergraph/cosmo/pull/1204/files#diff-e5f4b837be23c8cf9ed426c1bb3384f71ca32f1e63eba160ae024d5f99ad6220R2300

Unless I'm misunderstanding